DMTN-148.tex

```diff
@@ -50,9 +50,255 @@
+following this process.  These \verb|DEFECTS|-like calibration products have
+their origins as human-curated text files in the appropriate
+\verb|obs_CAMERA_data| packages (and are thereby versioned through the
+packages' git repositories).  The text file must be in a structured
+format, to allow the header keywords to be read and assigned to a
+metadata PropertyList.
+
+For these products to be used, they must be converted into a FITS
+format that can be read by ingestCalibs.py.  This is performed by the
+\verb|pipe_tasks|/ingestCuratedCalibs.py script, which uses the metadata
+information and the calibration product's python class to read the
+text version and convert to a FITS file format that can be ingested by
+ingestCalibs.py.
+
+This updated calibration process removes the necessity that each
+\verb|obs_CAMERA_data| package implement one-off mapper methods to allow the
+product to be found.  However, only \verb|DEFECTS|, \verb|QE_CURVE|, and \verb|LINEARITY|
+follow this system, leaving the remaining products to be handled with
+camera-specific code and cameraGeom entries.  In addition to having
+diverse implementations, these handlers are often rigid, and cannot be
+versioned easily.
+
+\section{Transitioning to Gen3/What We Should Expect From Gen3}
+
+\subsection{A CalibrationBaseClass}
+The first step to a consistent calibration system is to ensure that
+all products in both generations use the \verb|DEFECTS|-like calibration
+product system.  For FITS image products, this requires future
+calibrations to be constructed with all of the header keyword values
+listed above.
+
+The remaining unconverted calibration types (\verb|CROSSTALK|, \verb|BFKERNEL|)
+should be converted to use \verb|DEFECTS|-like class, most likely by defining
+a CalibrationBaseClass, and making all of the calibration products
+subclasses of that.  This would enforce that all calibrations can be
+relied upon to provide getMetadata, setMetadata, readFits, writeFits,
+readText, and writeText methods.  Gen2 Butler instances would then be
+able to access all calibration products in a single consistent manner.
+This also ensures that Gen2 and Gen3 can use the same calibration
+products.
+
+\subsection{Calibration Construction in Gen3}
+
+Calibration products should be constructed using well defined
+PipelineTask code.  This removes the need to delineate between
+``calibrations'' and ``user-curated calibrations''.  The only
+remaining difference between these two groups is what the initial
+persisted data type is: ``calibrations'' are stored in FITS images, as
+they represent corrections that are only meaningful as full images;
+``user-curated calibrations'' can be visualized in a small number of
+parameters, and so persisting these in a text format to allow user
+inspection is reasonable.  ECSV and YAML are good formats for these
+calibrations, and a calibration's readText/writeText methods should
+implement one of these.
```

Comment:
I have the impression that on DM-24225 we were moving away from storing the linearizer LUT in a plain text format.

Reply:
No. There was some confusion there over text vs binary use cases. We are still saying that we want text in git but the fast binary form should be in butler. That has never changed. Having a fast text form helps as well but is not as critical as what the butler is using.